[WIP] Update VCS Integration UX and rework EditorVCSInterface for network features

twaritwaikar opened the pull request:
This PR solves the conflicts in #39255, makes it merge ready, fixes a few bugs, and updates the EditorVCSInterface API.

This PR adds breaking changes in the VCS integration. VCS plugins made in the past will need heavy changes to comply with the newer API which supports features like push, pull, fetch, branch changes, etc.

The Git plugin is also being updated with this PR at the same pace (https://github.com/godotengine/godot-git-plugin/pull/84).

4.x will use GDExtension instead of GDNative, so I have only opened the PR against 3.x.

Janglee123 committed:
Updated commit dock UI

Janglee123 committed:
improved diff view

Janglee123 committed:
Replaced Dictionary/Array with custom structure

Janglee123 committed:
Added version control field in project settings
Now, version control editor plugin will auto load plugin on startup once it
initialized by user. Also this opens a window to configure attributes
of plugin and interface. :)

Janglee123 committed:
Added error popup window

Janglee123 committed:
Fixed opened scene not reloading after changes on disk

Janglee123 committed:
Updated doc

twaritwaikar committed:
Refactor error state logs in EditorVCSInterface

Also renames `addon` to `plugin` in the context of the VCS interface

Janglee123 committed:
Hide API methods from gdscript

Janglee123 committed:
Updated bindings

twaritwaikar committed:
Merge branch '3.x' into update-vcs

twaritwaikar committed:
Make EditorVCSInterface store less amount of internal state

twaritwaikar committed:
Add force push checkbox + more frequent VCS updates

Add force push checkbox in the Commit dock. Also add some missing
opportunities for checking the VCS state again on from UI inputs

ChronicallySerious commented:
I am attaching a few screenshots from a build to date.

A large part of this UI is thanks to @Janglee123's work in the past! 

Requesting any comments, suggestions because this PR is going to reach merge-quality soon enough.

![image](https://user-images.githubusercontent.com/31801364/136295577-2f948ad0-0065-4533-a1f7-da05c1d7ed6c.png)

Split/Unified diffs.
![image](https://user-images.githubusercontent.com/31801364/136296109-e43d91c3-7cc1-487c-b6e5-492d7dde4391.png)
![image](https://user-images.githubusercontent.com/31801364/136296146-84021213-2458-4a51-8857-372255eafe10.png)


The Editor stores the remote username and password, but it doesn't save them on the file system due to security issues. Could there be a better way where the editor is able to pick up the username and password but not a hacker?
![image](https://user-images.githubusercontent.com/31801364/136295659-78c6358e-c7a7-4330-a757-87cc2062606c.png)

SSH access needs testing. Access to repositories using a Github PAT token is tested.

ChronicallySerious commented:
However separating out the network related operations into different areas sounds a bit un-intuitive because the normal sentiment for seeing a pull button is to assume the push button is somewhere close

twaritwaikar committed:
Fix script contents not being updated on merge conflict

twaritwaikar committed:
Add branch creation VCS interface calls

twaritwaikar committed:
Add VCS remote creation and remote selection menus

twaritwaikar committed:
Show more commit information + Fix truncated commit offets

ChronicallySerious commented:
Now the Version Control dock and the Commit dock both display more commit information
![image](https://user-images.githubusercontent.com/31801364/136643088-56a26fe6-b3de-4248-a3d5-c2e6468c1c7d.png)

twaritwaikar committed:
Make VCS less noisy + Fix diff view refreshes

twaritwaikar committed:
Add support for SSH credentials for VCS remote auth

twaritwaikar committed:
Store VCS settings in Project and Editor Settings

Store username and SSH key paths in Editor Settings.
Shift the VCS autoload settings registrations to project_settings.cpp

ChronicallySerious commented:
Marking as ready for review ✔️ 
However, there are a couple of features I want to add to this like being able to set the commit list size from the UI and then writing documentation for the new functions.

But opening this for review because the main feature set is pretty much done.

twaritwaikar committed:
Do not strip VCS passwords with spaces when checking emptiness

twaritwaikar committed:
Add commit list size button to control commit history depth

twaritwaikar committed:
Add documentation for the new EditorVCSInterface

twaritwaikar committed:
Add options for deleting remotes and branches

ChronicallySerious commented:
Added branch and remote deletion UI. For this, I also added a MenuButton at the end of the network features toolbar and shifted all the less used features over there to reduce clutter in the main UI.

![image](https://user-images.githubusercontent.com/31801364/137565742-8aeee4f4-0deb-4b31-ab64-1410e6e6279c.png)
![image](https://user-images.githubusercontent.com/31801364/137565745-a1cde71e-e030-4dc0-a904-aa53d496e95c.png)

ChronicallySerious commented:
@fire Yes, I would agree that breaking it up makes the most sense. However, this PR is adding only those features which directly affect the `EditorVCSInterface` API, which needs to stay consistent in the different versions of the API, in other words, the API was revised once for 3.2, and the API is being revised the 2nd time in this PR.

If I were to break the changes to this API into different PRs, then there would be commits in the Godot history that would not have a fully complete revision of the API so the VCS plugins would break for those intermediate versions of Godot, which is where I would like to hear more opinions from other contributors.

ChronicallySerious commented:
In reply to my last comment:

I have squashed a few commits and will now close this PR. I will be opening new PRs with each new PR starting with a commit that directly changes the EditorVCSInterface API.

This way, the PR count for the same features shouldn't get out of control and still make it possible for this PR to be broken into smaller ones.

(I realized the cost of adding in code that is not reviewed or risking the work to be pending for long is not worth adding a few commits in history which don't work with the plugin that's currently in sync with the most recent version of EditorVCSInterface)

ChronicallySerious commented:
FYI: New PR with only the first batch of changes is up here: https://github.com/godotengine/godot/pull/53900

I will port any code changes that I make to address any review comments in the PR, and then rebase the next batch of commits on it, and then send another PR with those commits.
